scripts/firmware.py

```diff
@@ -0,0 +1,2097 @@
+            if choice == 0:
+                print(Utils.colored_text("Exiting...", Color.CYAN))
+                exit()
+
+            # Validate and handle the choice
+            if self.is_valid(choice):
+                self.execute(choice)
+                break
+            else:
+                self.invalid()
+
+    def get(self) -> int:
+        """Get and validate user input as an integer."""
+        try:
+            return int(
+                input(Utils.colored_text(" Select an option: ", Color.YELLOW)).strip()
+            )
+        except ValueError:
+            print(
+                Utils.colored_text("Invalid input. Please enter a number.", Color.RED)
+            )
+            return -1  # Return -1 for invalid input
+
+    def is_valid(self, choice: int) -> bool:
+        """Check if the user's choice is valid."""
+        return choice in self.menu_items and isinstance(
+            self.menu_items[choice], self.MenuItem
+        )
+
+    def execute(self, choice: int) -> None:
+        """Execute the action associated with a valid menu choice."""
+        menu_item = self.menu_items[choice]
+        if isinstance(menu_item, self.MenuItem):
+            menu_item.action()  # Safe to call because type is now narrowed
+        else:
+            raise TypeError(f"Expected MenuItem, got {type(menu_item).__name__}")
+
+    def invalid(self) -> None:
+        """Display a message for an invalid choice."""
+        Utils.header()
+        print(Utils.colored_text("Invalid choice. Please try again.", Color.RED))
+
+
+class Validator:
+    """A utility class for common validation checks."""
+
+    def __init__(self, firmware: "Firmware"):
+        self.firmware: Firmware = (
+            firmware  # Reference to the firmware object for navigation
+        )
+
+    def validate_device(self, device: str, type: str) -> bool:
+        if type == "CAN":
+            device_regex = r"^[a-f0-9]{12}$"
+        elif type == "USB":
+            device_regex = r".*Cartographer.*"
+        else:
+            device_regex = r"^[a-f0-9]{4}:[a-f0-9]{4}$"
+        return bool(re.match(device_regex, device))
+
+    def check_selected_firmware(self):
+        if not self.firmware.selected_firmware:
+            self._error_and_return("You have not selected a firmware file.")
+
+    def check_selected_device(self):
+        if not self.firmware.selected_device:
+            self._error_and_return("You have not selected a device to flash.")
+
+    def check_temp_directory(self):
+        if self.firmware.dir_path is None:
+            self._error_and_return("Error getting temporary directory path.")
+
+    def _error_and_return(self, message: str):
+        Utils.error_msg(message)
+        _ = input(
+            Utils.colored_text(
+                "\nPress Enter to return to the main menu...", Color.YELLOW
+            )
+        )
+        self.firmware.main_menu()
+
+
+class Firmware:
+    can: "Can"
+    usb: "Usb"
+    dfu: "Dfu"

```

Comment:
Should this be an enum?
```suggestion
class FlashMethod(StrEnum):
	CAN: auto()
	USB: auto()
	DFU: auto()

```